:rocket: Use macros from `limits.h` for bounds when checking for overflow

MisterDA opened the pull request:
The code is currently correct since we use wrap-around semantics for signed integers (`-fwrapv`), but:
- it's difficult to communicate that fact to static analyzers, which warn when computing the minimum integer with left-shifting 1 to the sign bit position (most-significant bit);
- MSVC doesn't support wrap-around semantics, but historically hasn't optimized for this, and might innocuously warn. 

From the standard (which I recall doesn't consider wrap-around semantics for signed integers):

> The result of `E1 << E2` is `E1` left-shifted `E2` bit positions; vacated bits are filled with zeros. […] If `E1` has a signed type and nonnegative value, and `E1 × 2^E2` is can't be represented in the result type, then that is the resulting value; otherwise, the behavior is undefined.

The problem being that the result of `1 << CHAR_BIT * sizeof(int) - 1` to compute the minimum `int` can't be represented in the result type (it's 2^63, but the maximum is 2^63-1); without wrap-around.

Using constants from `<limits.h>` instead allows for self-documenting code and silences these warnings.

Introduce the `INTNAT_MIN` macro to avoid independent re-definitions of this value.

gasche committed:
matching: specialized handling of the global 'fail exit'

This approach should preserve the current compilation behavior. It is
more efficient for [Partial] matches, as we avoid tracking any context
information on the final exit.

We hope that this refined interface will also let us implement a fine-grained
fix for totality failures due to mutable value. The idea is that when
we detect that some current values may flow to the final exit, we
could sometimes weaken [Jumps.empty Total] into [Jumps.empty Partial]
if we are under a mutable constructor.

gasche committed:
matching: debug the jumps partiality information

MisterDA committed:
Use macros from limits.h for bounds when checking for overflow

Introduce the macro INTNAT_MIN.

ghost commented:
Why not simply use `1U << 31` and `1ULL << 63` for the usual cases?

MisterDA commented:
> Why not simply use `1U << 31` and `1ULL << 63` for the usual cases?

This raises warnings:
```
comparison of integers of different signs: 'int64_t' (aka 'long long') and 'unsigned long long'
```

```c
CAMLprim value caml_int64_div(value v1, value v2)
{
  int64_t dividend = Int64_val(v1);
  int64_t divisor = Int64_val(v2);
  if (divisor == 0) caml_raise_zero_divide();
  /* PR#4740: on some processors, division crashes on overflow.
     Implement the same behavior as for type "int". */
  if (dividend == (1ULL << 63) && divisor == -1) return v1;
  return caml_copy_int64(dividend / divisor);
}
```

We could re-cast the expression to the `dividend` type, but at this stage isn't it easier to just use the `limits.h` macro?

ghost commented:
> We could re-cast the expression to the `dividend` type, but at this stage isn't it easier to just use the `limits.h` macro?

Sigh. Yes, at this point, your changes make more sense.

tmcgilchrist committed:
Add specific instructions on removing a primitive.

MisterDA committed:
Use macros from limits.h when serializing ints and nativeints

This fixes the warning from MSVC raised on -0x80000000.

> warning C4146: unary minus operator applied to unsigned type, result
> still unsigned

The other replacements are made for consistency and, hopefully,
legibility.

gasche committed:
matching: massage debug output to get rid of {Def,Jumps}.is_empty

gasche committed:
Changes

gasche committed:
Merge pull request #13076 from gasche/matching-bug-final-exit

The Pattern-Matching Bug: special handling of final exits

gasche committed:
Merge pull request #13064 from tmcgilchrist/bootstrap_docs

Add specific instructions on removing a primitive.
